rootconf/grub-arch/sysroot-lib-onie/exec-installer-arch

```diff
@@ -12,6 +12,29 @@ max_log_lines=1000
 # Maximum number of attempts to install a firmware update.
 max_fw_attempts=5
 
+# This function is called from /bin/exec_installer at the completion
+# of the NOS install operation.
+finish_nos_install()
+{
+    local result="$1"
+    local URL="$2"
+    local update_image="$3"
+
+    if [ $result -eq 0 ] ; then
+        log_console_msg "NOS install successful: $URL"
+
+        # Set ONIE boot mode to 'none' after a successful install.
+        onie-boot-mode -q -o none
+
+        # Set the NOS mode to 'yes' after a successful install.
+        onie-boot-mode -q -n yes
```

Comment:
seems there are a couple of ways to get there.

one possibility:
- have the diag run `onie-boot-mode -n no` whenever it boots, thereby clearing the NOS install mode.  unfortunately that does not help if the diag is only installed and never booted.

another possibility:
- have the diag installer leave a marker that the ONIE install driver can inspect.  One example the diag installer touches `/var/run/diag-install` and then the exec-installer code looks for that file.  If that file exists then do not set the NOS install mode.  Another example is to have the diag installer exit with a special exit code.  Again the exec-installer code could inspect the exit code and say "yes, this is a successful diag install.  skip setting NOS mode".

Either way the diag runtime or the diag installer would need to change a little.

I like the later alternative, where the diag installer passes a bit of info.  I'm not too particular about touching a file or using an exit code.

thoughts?

Reply:
> have the diag run `onie-boot-mode -n no` whenever it boots,

This seems to be not feasible.  If a machine has installed both NOS and diag already, booting to diag would cause NOS mode cleared.

> have the diag installer leave a marker that the ONIE install driver can inspect.

This is a way to reach the goal to distinguish diag from NOS.  It needs to revise present installers to be aware of the new way.  However, there are many old installers living in the world for mass production.  Old installer executed in new ONIE will turn on NOS mode.  And most time a machine would have NOS.  So a machine should almost stay in NOS mode.  Eventually people may always enable NOS mode to make install mode sticky lose efficacy.

Let's back to check the aim of install mode sticky.  In commit acf96ce, it intends to solve the problem:

> The problem was the 'install' and 'update' states were not persistent
across reboots.  Imagine entering those states and then the power goes
off.  Upon power-up the state would be lost and the system would boot
into whatever was previously configured.

I would share you another thought for the problem to replace current approach of making install mode sticky.

 For u-boot arch, it would always boot to install mode if `onie_boot_reason` is not cleared.  The boot reason can be cleared by `finish_nos_install()` when NOS installer doesn't report any error.

For grub arch, we can use `50_onie_grub` to generate code for NOS grub.cfg to load specific file (such as /mnt/onie-boot/grub/onie_boot_reason) to know current boot reason.  If the boot reason is one of 'install', 'rescue', 'update', 'uninstall' or 'embed', the code would perform `set default=ONIE` to make NOS grub menu to select ONIE by default:

    +----------------------------------+
    | DEMO OS                          |
    |*ONIE                             |  <-- default
    |                                  |
    |                                  |

The boot reason is set/reset by `onie-boot-mode`.  For instance, when `onie-boot-mode -q -o install` is run, it sets `onie_mode` and update /mnt/onie-boot/grub/onie_boot_reason at the same time.  NOS grub will know the fresh boot reason when it loads the file.

This approach would be more flexible than present one.  But it has a bit of flaw in grub arch.  If a installer reinstall NOS failed and tell ONIE it finished successfully to reboot, it will go to a broken NOS grub.  To prevent the false report, it needs to make ONIE sticky (set to boot ONIE grub) one time before running NOS installer in *install mode*.

Reply:
> This seems to be not feasible. If a machine has installed both NOS and diag already, booting to diag would cause NOS mode cleared.

I agree, good point.  That method will not work.

Reply:
Another thought - instead of ONIE automatically setting the NOS mode in `finish_nos_install()`, make setting NOS mode the responsibility of the NOS installer.  So the NOS installer must set NOS mode to true.

For a diag installer, no changes are necessary.  A diag installer would not set NOS mode to true.  Install mode would still be sticky.

A NOS installer would set NOS mode to true.  Of course this would require a change to NOS installers.  I'll go through the case below to see the gotchas:

Old ONIE, Old installer -- no change in current behavior.  install mode is sticky.  This is where we are today.

Old ONIE, new installer -- the `onie-boot-mode` -n option will not be available.  The NOS installer should test to see if it works and discard any errors.  Net result -- no change in current behavior.  install mode is sticky.

New ONIE, old installer -- no change in current behavior.  install mode is sticky.

New ONIE, new installer -- the `onie-boot-mode` -n option *is* available.  install mode is no longer sticky.

To me that is not so bad.  In all cases we are no worse than before and once NOS installers are updated we have a much better behavior.

It does feel "more correct" to make the setting of NOS mode the responsibility of the NOS installer, that way ONIE does not have to guess.

Thoughts?

Reply:
@cbrune ,

Everything looks good to me.

It is nice to have the feature in `run_installer()` like:

    # install mode sticky before executing installer
    if [ "$onie_boot_reason" = "install" ] &&
       [ ! -e "/var/run/onie/install-mode-sticky" ] ; then
        install_remain_sticky_arch
    fi

    # start executing installer
    ...

Reply:
OK, I think we are converging -- We will make it the responsibility of the NOS installer to set the NOS mode.

As far as "touching" a file, I think it best to hide all that detail behind a command that lives in `/bin`, something like `onie-set-nos-mode`.  That way both u-boot and grub architectures can call it without knowing the details.  The different architecture backends can provide a particular implementation.  Also easier to document a command, rather than documenting a rather cryptic "touch a file" concept.

Also will modify the demo OS to use the new technique, illustrating its usage.

Regarding the install sticky before running the installer -- can we leave that off for now?  That is solving an important, but separate issue.  I'd like to get the NOS mode finalized first.

Reply:
pls update the documentation section for NOS and diag installer, so that the folks implementing the NOS/diag installers can benefit from this feature.

Reply:
> Regarding the install sticky before running the installer -- can we leave that off for now? That is solving an important, but separate issue. I'd like to get the NOS mode finalized first.

No problem.  Thank you very much.